iommu.c: rework initialization function

krystian-hebel opened the pull request:
Test for INVALIDATE_ALL completion is now done inside the initialization function, instead of main.c. Function names were changed to better describe what the function does. Comment describing window in anti-DMA security was updated with new finds.

Two bugs were fixed:
- EventLogInt wasn't properly cleared (it is write-1-to-clear bit)
- CmdBufEn was set on transition to the kernel, which didn't clear it before setting up head/tail pointers, which in turn lead to hang (it is listed as undefined behavior in IOMMU specification)

krystian-hebel committed:
iommu.c: rework initialization function

Test for INVALIDATE_ALL completion is now done inside the initialization
function, instead of main.c. Function names were changed to better
describe what the function does. Comment describing window in anti-DMA
security was updated with new finds.

Some bugs were fixed:
- EventLogInt wasn't properly cleared (it is write-1-to-clear bit)
- CmdBufEn was set on transition to the kernel, which didn't clear it
  before setting up head/tail pointers, which in turn lead to hang
  (it is listed as undefined behavior in IOMMU specification)
- Command buffer tail/head pointer registers were incorrectly masked

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk commented on iommu.c:
```diff
@@ -122,7 +130,7 @@ u32 iommu_load_device_table(u32 cap, volatile u64 *completed)
 
     /* Clear EventLogInt set by IOMMU not being able to read command buffer */
-    mmio_base[IOMMU_MMIO_STATUS_REGISTER] &= ~2;
+    mmio_base[IOMMU_MMIO_STATUS_REGISTER] &= IOMMU_SR_EventLogInt;
```
`~` got lost.

krystian-hebel replied:
No, it actually shouldn't be there in the first place. This bit is write-1-to-clear.

SergiiDmytruk replied:
Maybe update comment to say that it's write-1-to-clear?  I read about it in PR's description, but some time later saw "clear" and `&=` without `~` and assumed the code is wrong.

krystian-hebel replied:
https://github.com/TrenchBoot/secure-kernel-loader/compare/43049429092897dd3fe6663eb9c5329f12f012b1..d2cb1b776cd1c44abedbd0f68401d62c46b6e05a

SergiiDmytruk commented on iommu.c:
```diff
@@ -36,17 +36,28 @@ static void send_command(u64 *mmio_base, iommu_command_t cmd, u32 *idx)
     smp_wmb();
     mmio_base[IOMMU_MMIO_COMMAND_BUF_TAIL] =
-        _u(&command_buf[*idx]) - (_u(&command_buf) & ~0xfff);
+        _u(&command_buf[*idx]) - (_u(&command_buf) & ~0x1fff);
```
Why change masks from 4KiB to 8KiB alignment here and in `iommu_init()`?  If the whole point of advertising command buffer to be 8KiB is to make sure that aligning it to 4KiB boundary won't leave end of `command_buf` beyond `base address + 4KiB`, isn't aligning it to 8KiB with the size of 8KiB allows for the same problematic situation to occur?

krystian-hebel replied:
You are correct, I misread how this is parenthesized and thought that whole tail register value is masked. Undone in https://github.com/TrenchBoot/secure-kernel-loader/compare/d2cb1b776cd1c44abedbd0f68401d62c46b6e05a..f16a0d0651f8dc5a22d098ee98f6f9daf4d98618

krystian-hebel committed:
tpmlib/tis.c: work around poor TPM implementations

Release versions of SKL hanged in tis_send(), while versions with serial
output worked, most likely thanks to increased delay between DRTM
sequence and next TPM command. Apparently, some TPMs don't properly
handle setting STS.commandReady when all of the following conditions are
met:

- TPM has recently finished DRTM sequence (internal work may still be
  happening at that point, there is no way to be sure),
- TPM just transitioned to Idle state (e.g. by changing locality),
- STS.commandReady is written periodically before TPM reports it is in
  Ready state.

When all of the above applies, STS.commandReady is always read as 0,
as if the TPM restarted transition from Idle to Ready each time it is
asked to do so. To work around this, set this bit once and keep checking
in a loop until TIMEOUT_B (2 seconds). Well behaving TPM must be able to
enter Ready state before that time, if it doesn't, error is returned.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
SNAFU. Binary built locally from this code (on old Ubuntu, will check newer Fedora) works, while binary built by CI reports wrong self-measurement hash. Other hash values seem to be valid (PCR replay gives current value after fixing just the first entry), so the problem most likely lies in `_end_of_measured`. I think this might be caused by `-R '.note.*'`.

EDIT: shouldn't be, `_end_of_measured` has the same address as base of `.note.gnu.property`:

```
[khebel@3M05 secure-kernel-loader]$ objdump skl -h

skl:     file format elf64-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         000029a0  0000000000000000  0000000000000000  00001000  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  1 .rodata       00000270  00000000000029a0  00000000000029a0  000039a0  2**5
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .note.gnu.property 00000030  0000000000002c10  0000000000002c10  00003c10  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  3 .bss          0000a4d4  0000000000003000  0000000000003000  00004000  2**12
                  ALLOC
  4 .comment      0000002e  0000000000000000  0000000000000000  00003c40  2**0
                  CONTENTS, READONLY
  ( ... more debug sections ... )
[khebel@3M05 secure-kernel-loader]$ objdump skl -t | grep _end_of_measured
0000000000002c10 g       .rodata	0000000000000000 .hidden _end_of_measured
```

krystian-hebel commented:
Local build on Fedora also doesn't give proper hash...

krystian-hebel committed:
.github/workflows/build.yml: run 'make tests'

This enables SHA tests to check if different compilers produce the same
results.

Definitions in string.h were moved outside of __STDC_HOSTED__ to avoid
compiler warnings (promoted to errors because of -Werror) due to
incompatible implicit declaration of built-in functions.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
Something changed in gcc-11, it fails to produce valid SHA1 measurements: https://github.com/TrenchBoot/secure-kernel-loader/pull/19

krystian-hebel committed:
sha1sum.c, sha256.c: make code work with -fstrict-aliasing

-fstrict-aliasing is enabled by default when using optimization levels
higher than 1, including -Os. With that option, compiler may assume
that object of one type never resides at the same address as object of
a different type. Both sha1_final() and sha256_final() used to write
message length by casting a pointer to buffer into a pointer to u64,
while surrounding code operated on the buffer directly.

The problem manifests in GCC 11 and later versions.

The commit fixes this issue and another UB caused by unaligned access
at the beginning of transformation step by using memcpy() in both cases.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
